compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -2332,13 +2333,134 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         self.ast_ty_to_ty_inner(ast_ty, false, true)
     }
 
+    fn check_delegation_for_opaque_types(&self, sig: ty::PolyFnSig<'tcx>) -> bool {
+        struct OpaqueSeearchVisitor;
+
+        impl<'tcx> ty::visit::TypeVisitor<TyCtxt<'tcx>> for OpaqueSeearchVisitor {
+            type BreakTy = bool;
+
+            fn visit_ty(&mut self, ty: Ty<'tcx>) -> ControlFlow<Self::BreakTy> {
+                if let ty::Alias(alias_kind, _) = ty.kind()
+                    && let ty::AliasKind::Opaque = alias_kind
+                {
+                    return ControlFlow::Break(true);
+                }
+                ty.super_visit_with(self)
+            }
+        }
+
+        OpaqueSeearchVisitor.visit_binder(&sig).break_value().unwrap_or_default()
+    }
+
+    fn check_delegation_constraints(&self, res_id: DefId, span: Span, emit: bool) -> bool {
+        #[inline]
+        fn generics_param_count(param_count: ty::GenericParamCount) -> usize {
+            param_count.lifetimes + param_count.types + param_count.consts
+        }
+
+        let mut error_occured = false;
+        let res_span = self.tcx().def_span(res_id);
+        let mut try_emit = |descr| {
+            if emit {
+                self.tcx().dcx().emit_err(crate::errors::NotSupportedDelegation {
+                    span,
+                    descr,
+                    res_span,
+                });
+            }
+            error_occured = true;
+        };
+
+        if let Some(node) = self.tcx().hir().get_if_local(res_id)
+            && let Some(decl) = node.fn_decl()
+            && let hir::FnRetTy::Return(ty) = decl.output
+            && let hir::TyKind::InferDelegation(_, _) = ty.kind
+        {
+            try_emit("recursive delegation");
+        }
+
+        let res_sig = self.tcx().fn_sig(res_id).instantiate_identity();
+        if self.check_delegation_for_opaque_types(res_sig) {
```

Comment:
```suggestion
        if res_sig.output().has_opaque_types() {
```
Something like this should be enough, I think.
(Types in function parameters cannot define opaque types, and using opaque types defined elsewhere should be fine.)